pkg/kubelet/types/pod_status.go

```diff
@@ -42,5 +42,10 @@ func PodConditionByKubelet(conditionType v1.PodConditionType) bool {
 			return true
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+		if conditionType == v1.AlphaNoCompatGuaranteeDisruptionTarget {
```

Comment:
is this condition owned by the kubelet? other actors can set it as well, right? do we need to pass in the reason as well and see if the reason is `TerminationByKubelet`?

Reply:
Yes, kubelet owns the condition, but it will not drop the condition added by other actors thanks to this check:
https://github.com/kubernetes/kubernetes/blob/4297fa59d02b9b92887f120abf15468fc82489d5/pkg/kubelet/status/status_manager.go#L884
if kubelet does not have a new condition it just uses the one from API server.

Reply:
It is also not a precedence that a pod condition owned by kubelet is also updated by other components - for example here scheduler updates the `PodScheduled` condition: https://github.com/kubernetes/kubernetes/blob/b1e348f60debac7455ee2d0a740d90101c0ad171/pkg/scheduler/schedule_one.go#L909. On the other hand, if the condition was not owned by kubelet, then the internal cache in kubelet might not be updated due to this check: 
https://github.com/kubernetes/kubernetes/blob/b1e348f60debac7455ee2d0a740d90101c0ad171/pkg/kubelet/status/status_manager.go#L549

Reply:
hmm... special-casing the merge of this one condition that makes this function *really* hard to follow.

Would it be clearer to add the concept of "shared conditions" that the kubelet preserves if it doesn't have an opinion, but overrides if it does:

```diff
        podConditions := make([]v1.PodCondition, 0, len(oldPodStatus.Conditions)+len(newPodStatus.Conditions))
 
        for _, c := range oldPodStatus.Conditions {
                if !kubetypes.PodConditionByKubelet(c.Type) {
                        podConditions = append(podConditions, c)
                }
        }
        for _, c := range newPodStatus.Conditions {
                if kubetypes.PodConditionByKubelet(c.Type) {
                        podConditions = append(podConditions, c)
+               } else if kubetypes.PodConditionSharedByKubelet(c.Type) {
+                       // update or append in podConditions
+                       if i, _ := podutil.GetPodConditionFromList(podConditions, c.Type); i >= 0 {
+                               podConditions[i] = c
+                       } else {
+                               podConditions = append(podConditions, c)
+                       }
                }
        }
        newPodStatus.Conditions = podConditions
```

Reply:
Seems like a sensible idea. What do you think @bobbypage ?

Reply:
@liggitt pushed a commit which introduces the concept of shared pod condition types to simplify the function. PTAL

Reply:
+1, I like the idea of "shared pod condition", it definitely helps to simplify the logic here.